### ISSUE TITLE
Type constraints are not enforced in guard function return assertion

### ISSUE THREAD

**Issue by gcnew** [Microsoft/TypeScript#17297]
**TypeScript Version:**  nightly (2.5.0-dev.20170629)

**Code**

```ts
type Keys = 'a'|'b'|'c'
type KeySet<T extends Keys> = { [k in T]: true }

// expected an error, since Keys doesn't have a 'd'
declare function hasKey<T extends Keys>(x: KeySet<T>): x is KeySet<T|'d'>;
```

**Expected behavior:**
The above `hasKey` function should be an error, since it adds `'d'` to the `KeySet`, however `KeySet` has a constraint that `T` can only be a subset of `Keys`.

**Actual behavior:**
`hasKey` compiles just fine.

**Comment by jcalz** [comment]
Yeah, this seems buggy.  It has nothing to do with the fact that `hasKey()` is generic, in case anyone cares:

```ts
type Foo = { 'a': string; }
type Bar = { 'a': number; }
type NeedsFoo<T extends Foo> = { foo: T };
declare var anError: NeedsFoo<Bar>; // error, as expected
declare var alsoAnError: NeedsFoo<number>; // also error, as expected
declare function noError(x: any): x is NeedsFoo<Bar>; // no error!
declare function alsoNoError(x: any): x is NeedsFoo<number>; // also no error!
```

**Comment by RyanCavanaugh** [comment]
`checkTypePredicate` needs to call `checkSourceElement(node.type)`
